tests/ArrayAssignmentTest.php

```diff
@@ -1543,8 +1648,113 @@ function test(): array {
 
                     $x = [...test(), "a" => "b"];
                 ',
-                'assertions' => ['$x' => 'non-empty-array<int|string, mixed|string>']
+                'assertions' => ['$x===' => "array{a: 'b'}<int<0, max>, mixed>"],
             ],
+            'checkTraversableUnpackTemplatesCorrectly' => [
+                'code' => '<?php
+                    /**
+                     * @template T1
+                     * @template T2
+                     * @template TKey
+                     * @template TValue
+                     * @extends Traversable<TKey, TValue>
+                     */
+                    interface Foo extends Traversable {}
+
+                    /**
+                     * @param Foo<"a"|"b", "c"|"d", "e"|"f", "g"|"h"> $foo
+                     * @return array<"e"|"f", "g"|"h">
+                     */
+                    function foobar(Foo $foo): array
+                    {
+                        return [...$foo];
+                    }
+                ',
+                'assertions' => [],
+                'ignored_issues' => [],
+                'php_version' => '8.1',
+            ],
+            'unpackIncorrectlyExtendedTraversable' => [
+                'code' => '<?php
+                    /** @extends Traversable<int> */
+                    interface Foo extends Traversable {}
+
+                    /**
+                     * @return array<int, mixed>
+                     */
+                    function foobar(Foo $foo): array
+                    {
+                        return [...$foo];
+                    }
+                ',
+            ],
```

Comment:
I think I meant to go back and look at how this should work, it seems wrong doesn't it? I know we allow omitting template params for arrays, but I'm not sure if we do for traversable. If we do though, I would think it would unpack into `array<array-key, int>` instead of `array<int, mixed>`.

Reply:
That's only about forcing templates to be explicitly defined, this is about `Traversable` allowing either 1 or 2 templates so that you can omit the key, and this doesn't handle it correctly since it's using the single template as the key instead of as the value.

Reply:
Turns out I accidentally used `canBeContainedBy` instead of `isContainedBy`, it actually was being resolved as `list<int>` which is a subtype of `array<int, mixed>` so it went unnoticed.